File: autocontext/src/autocontext/execution/judge_executor.py
from __future__ import annotations

from autocontext.scenarios.agent_task import AgentTaskInterface, AgentTaskResult


class JudgeExecutor:
    """Executes evaluation by delegating to an AgentTaskInterface."""

    def __init__(self, task: AgentTaskInterface) -> None:
        self.task = task

    def execute(
        self,
        agent_output: str,
        state: dict,
        reference_context: str | None = None,
        required_concepts: list[str] | None = None,
        calibration_examples: list[dict] | None = None,
        pinned_dimensions: list[str] | None = None,
    ) -> AgentTaskResult:
        """Evaluate agent output using the task's evaluate_output method."""
        # Run context preparation if the task supports it
        prepared_state = self.task.prepare_context(dict(state))
        context_errors = self.task.validate_context(prepared_state)
        if context_errors:
            return AgentTaskResult(
                score=0.0,
                reasoning=f"Context validation failed: {'; '.join(context_errors)}",
                dimension_scores={},
            )

        return self.task.evaluate_output(
            agent_output,
            prepared_state,
            reference_context=reference_context,
            required_concepts=required_concepts,
            calibration_examples=calibration_examples,
            pinned_dimensions=pinned_dimensions,
        )

File: ts/src/execution/judge-executor.ts
/**
 * JudgeExecutor — evaluates agent output by delegating to AgentTaskInterface.
 * Port of autocontext/src/autocontext/execution/judge_executor.py
 */

import type { AgentTaskInterface, AgentTaskResult } from "../types/index.js";

export class JudgeExecutor {
  private task: AgentTaskInterface;

  constructor(task: AgentTaskInterface) {
    this.task = task;
  }

  /**
   * Evaluate agent output using the task's evaluateOutput method.
   * Runs context preparation and validation before judging.
   */
  async execute(
    agentOutput: string,
    state: Record<string, unknown>,
    opts?: {
      referenceContext?: string;
      requiredConcepts?: string[];
      calibrationExamples?: Array<Record<string, unknown>>;
      pinnedDimensions?: string[];
    },
  ): Promise<AgentTaskResult> {
    // Run context preparation if the task supports it
    const preparedState = this.task.prepareContext
      ? await this.task.prepareContext({ ...state })
      : { ...state };

    // Validate context
    const contextErrors = this.task.validateContext
      ? this.task.validateContext(preparedState)
      : [];

    if (contextErrors.length > 0) {
      return {
        score: 0.0,
        reasoning: `Context validation failed: ${contextErrors.join("; ")}`,
        dimensionScores: {},
        internalRetries: 0,
      };
    }

    return this.task.evaluateOutput(agentOutput, preparedState, opts);
  }
}
